pkg/cache/stats_counter.go

```diff
@@ -0,0 +1,229 @@
+)
+
+// StatsCounter is an interface for recording cache stats for goburrow.Cache.
+// The StatsCounter can be found at
+// - https://github.com/goburrow/cache/blob/f6da914dd6e3546dffa8802919dbca80cd33abe3/stats.go#L67
+var _ burrow.StatsCounter = (*burrowStatsCounter)(nil)
+
+// burrowStatsCounter is a StatsCounter implementation for burrow cache.
+// It is a wrapper around prometheus metrics.
+// It has been intended to passed through the cache using cache.WithStatsCounter option
+// - https://github.com/goburrow/cache/blob/f6da914dd6e3546dffa8802919dbca80cd33abe3/local.go#L552
+type burrowStatsCounter struct {
+	logger log.Logger
+	reg    prometheus.Registerer
+
+	hits     prometheus.Counter
+	miss     prometheus.Counter
+	eviction prometheus.Counter
+
+	trackLoadingCacheStats bool
+	loadSuccess            prometheus.Counter
+	loadError              prometheus.Counter
+	loadTotalTime          prometheus.Histogram
+}
+
+// Option add options for default Cache.
+type Option func(c *burrowStatsCounter)
+
+// WithTrackLoadingCacheStats enables tracking of loading cache stats.
+// It is disabled by default.
+func WithTrackLoadingCacheStats() Option {
+	return func(c *burrowStatsCounter) {
+		c.trackLoadingCacheStats = true
+		c.loadSuccess = promauto.With(c.reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_load_success_total",
+			Help: "Total number of successful cache loads.",
+		})
+		c.loadError = promauto.With(c.reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_load_error_total",
+			Help: "Total number of cache load errors.",
+		})
+		c.loadTotalTime = promauto.With(c.reg).NewHistogram(prometheus.HistogramOpts{
+			Name:    "cache_load_duration_seconds",
+			Help:    "Total time spent loading cache.",
+			Buckets: []float64{0.001, 0.002, 0.005, 0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1},
+		})
+	}
+}
+
+// NewBurrowStatsCounter creates a new StatsCounter.
+//
+// RecordLoadSuccess and RecordLoadError methods are called by Get methods on a successful and failed load respectively.
+// Get method only called by LoadingCache implementation.
+func NewBurrowStatsCounter(logger log.Logger, reg prometheus.Registerer, name string, options ...Option) *burrowStatsCounter {
+	reg = prometheus.WrapRegistererWith(prometheus.Labels{"cache": name}, reg)
+	s := &burrowStatsCounter{
+		logger: logger,
+		reg:    reg,
+
+		hits: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_hits_total",
+			Help: "Total number of cache hits.",
+		}),
+		miss: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_miss_total",
+			Help: "Total number of cache misses.",
+		}),
+		eviction: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_evictions_total",
+			Help: "Total number of cache evictions.",
+		}),
+	}
+	for _, opt := range options {
+		opt(s)
+	}
+	return s
+}
+
+// RecordHits records the number of hits.
+// This method is called by Get and GetIfPresent methods on a cache hit.
```

Comment:
Can you add explicitly to each function "RecordHits is part of the burrow.StatsCounter interface."? Otherwise I have to catch the link at the very start, remember the whole interface definition and go back here to understand.